File: solutions/free-pascal/kindergarten-garden/2/KindergartenGarden.pas
unit KindergartenGarden;

{$mode ObjFPC}{$H+}

interface

type
  Plant = (Grass, Clover, Radishes, Violets);

  TPlantArray = Array Of Plant;

function plants(const diagram, student : String) : TPlantArray;

implementation

uses SysUtils;

function letterToPlant(letter : char) : Plant;
begin
  case letter of 
    'G' : result := Grass;
    'C' : result := Clover;
    'R' : result := Radishes
    else
      result:=Violets;
  end;
end;

function plants(const diagram, student : String) : TPlantArray;
var
  stdNo, offset : integer;
begin
  result:=[];
  stdNo := ord(student[1]) - ord('A');
  offset := (length(diagram) div 2) + 2 + 2*stdNo;
  result:= TPlantArray.create(letterToPlant(diagram[2*stdNo+1]), 
                  letterToPlant(diagram[2*stdNo+2]), 
                  letterToPlant(diagram[offset]), 
                  letterToPlant(diagram[offset+1])); 

end;

end.

File: solutions/dart/kindergarten-garden/2/lib/kindergarten_garden.dart
enum Plant {
  radishes,
  clover,
  violets,
  grass,
}

enum Student {
  Alice,
  Bob,
  Charlie,
  David,
  Eve,
  Fred,
  Ginny,
  Harriet,
  Ileana,
  Joseph,
  Kincaid,
  Larry,
}

class KindergartenGarden {
  String diagram;

  KindergartenGarden(this.diagram);

  List<Plant> string2Plants(String patch){  
    return patch.split('').map((p)=>
      switch(p){
        'R' => Plant.radishes,
        'V' => Plant.violets,
        'C' => Plant.clover,
        'G' => Plant.grass,
        _ => throw ArgumentError('Unknown plant'),
      }).toList();
  }

  List<Plant> plants(Student child){
    var stNum = child.index;
    var offset =  (diagram.length ~/2) + 2*stNum +1;
    return string2Plants(
      diagram.substring(2*stNum, 2*stNum+2) +
    diagram.substring(offset, offset+2));
  }
}
